File: sections/21_graphs_i_mixed_contest/problems/g_alternating_road_route/solution.py
import heapq
import sys


def main():
    tokens = sys.stdin.buffer.read().split()
    n, m = map(int, tokens[:2])
    graph = [[] for _ in range(n)]
    index = 2
    for _ in range(m):
        u = int(tokens[index]) - 1
        v = int(tokens[index + 1]) - 1
        cost = int(tokens[index + 2])
        color = 0 if tokens[index + 3] == b"R" else 1
        index += 4
        graph[u].append((v, cost, color))
    if n == 1:
        print(0)
        return
    inf = 10**30
    dist = [[inf, inf] for _ in range(n)]
    heap = []
    for v, cost, color in graph[0]:
        if cost < dist[v][color]:
            dist[v][color] = cost
            heapq.heappush(heap, (cost, v, color))
    while heap:
        cost, u, last = heapq.heappop(heap)
        if cost != dist[u][last]:
            continue
        for v, weight, color in graph[u]:
            if color == last:
                continue
            new = cost + weight
            if new < dist[v][color]:
                dist[v][color] = new
                heapq.heappush(heap, (new, v, color))
    answer = min(dist[-1])
    print(-1 if answer == inf else answer)


if __name__ == "__main__":
    main()

File: sections/21_graphs_i_mixed_contest/problems/g_alternating_road_route/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, m;
    cin >> n >> m;
    struct Edge { int to, color; ll cost; };
    vector<vector<Edge>> graph(n);
    while (m--) {
        int u, v;
        ll cost;
        char color;
        cin >> u >> v >> cost >> color;
        graph[u - 1].push_back({v - 1, color == 'R' ? 0 : 1, cost});
    }
    const ll INF = (1LL << 62);
    vector<array<ll, 2>> dist(n, {INF, INF});
    using State = tuple<ll, int, int>;
    priority_queue<State, vector<State>, greater<State>> heap;
    for (int color = 0; color < 2; ++color) {
        for (auto edge : graph[0]) if (edge.color == color) {
            if (edge.cost < dist[edge.to][color]) {
                dist[edge.to][color] = edge.cost;
                heap.push({edge.cost, edge.to, color});
            }
        }
    }
    if (n == 1) {
        cout << 0 << '\n';
        return 0;
    }
    while (!heap.empty()) {
        auto [cost, u, last] = heap.top();
        heap.pop();
        if (cost != dist[u][last]) continue;
        for (auto edge : graph[u]) if (edge.color != last) {
            ll next = cost + edge.cost;
            if (next < dist[edge.to][edge.color]) {
                dist[edge.to][edge.color] = next;
                heap.push({next, edge.to, edge.color});
            }
        }
    }
    ll answer = min(dist[n - 1][0], dist[n - 1][1]);
    cout << (answer == INF ? -1 : answer) << '\n';
}
